added type hints

tim25651 opened the pull request:
Since the minimum supported version py3.8 supports type hints, type hints should be included where possible.
All tests succesfully for py3.8-py3.11
PR to master as no up to date dev branch is available.

Commit:
added type hints

Commit:
removed trailing whiteline in test file, and adjusted hashes in tests

Commit:
safely close archive after use

Commit:
contextmanager already implemented in arpy, using the implemented one

Commit:
do not repeat the logging, just pass the archive type

Commit:
made the revision string comparison type hinted and replaced the try-except indexerror with a zip_longest plus null check

memory commented on tests/testdeb_0.0.0.dsc:
```diff
@@ -18,4 +18,3 @@ Checksums-Sha256:
 Files:
  142ca7334ed1f70302b4504566e0c233 280 testdeb_0.0.0.orig.tar.gz
  fc80e6e7f1c1a08b78a674aaee6c1548 232 testdeb_0.0.0-1.debian.tar.xz
-
```
please undo this change, which will allow you to undo the changes to the checksums in test_dsc.py

memory reviewed (changes requested):
Thank you so much for this; it's long overdue and well appreciated.

Two requests:

1. Please revert the removal of the final EOL from testdeb_0.0.0.dsc, which in turn will allow you to revert the checksum changes in test_dsc.py
2. Please add a mypy stage to the test harness (.github/workflows/pr.yaml) and ensure it passes

Commit:
adjusted dstringcmp matching to compare_revision_strings, adjusted type hints, ignoring explicit override

Commit:
added mypy to CI

Commit:
reverted hash changes

caveats: in this change a linter which removes trailing whitespace needs to ignore this file,
else it will lead to test failure

ghost commented:
requested changes should be implemented.

memory reviewed (approved):
Thank you again for this!
